RealmAuditLog: Add RealmAuditLog entries whenever a CustomProfileFied is changed.

RajVishh opened the pull request:
Implements RealmAuditLog entries for custom profile field changes. Previously, creating fields or updating user profile data did not leave an audit trail.


Fixes #36715.

<details>
<summary>Self-review checklist</summary>

- [x] Self-reviewed the changes for clarity and maintainability
(variable names, readability, and consistency with the style guide).
- [x] Verified that the new test class works as expected.
- [x] Commit message explains reasoning and motivation for changes.
- [x] Each commit represents a coherent, reviewable change.
- [x] Made sure that the PR follows Zulip’s coding style.

Testing:-
- [x] Verified that the PR passes all the CI build tests.
</details>

RajVishh committed:
RealmAuditLog: Add RealmAuditLog entries whenever a CustomProfileField is changed.

Fixes #36715.

timabbott commented on zerver/actions/custom_profile_fields.py:
```diff
@@ -76,17 +85,42 @@ def try_add_realm_custom_profile_field(
+        event_type=AuditLogEventType.REALM_PROPERTY_CHANGED,  # Or specific type if available
+        event_time=now(),
+        extra_data=orjson.dumps(
+            {"property": "custom_profile_fields", "added": name, "type": field_type}
```
This is not how `AuditLogEventType.REALM_PROPERTY_CHANGED` extra data works.

timabbott commented:
@RajVishh I did a quick look at this, but @sahil839 will be the main reviewer for this. Please make sure the audit log `extra_data` is consistently in the standard formats.

RajVishh commented:
made the changes , please check @sahil839

sahil839 commented on zerver/tests/test_audit_log.py:
```diff
@@ -551,6 +540,106 @@ def test_set_realm_authentication_methods(self) -> None:
+
+        self.assertEqual(audit_log.acting_user, admin)
+        self.assertEqual(audit_log.extra_data["field_id"], field_id)
+
```
You should also add tests for user's value changing.

sahil839 commented on zerver/actions/custom_profile_fields.py:
```diff
@@ -76,17 +86,49 @@ def try_add_realm_custom_profile_field(
+        extra_data={
+            "field_id": custom_profile_field.id,
+            "name": name,
+            "field_type": field_type,
```
I am not sure if we need all the fields here for the custom profile.

sahil839 commented on zerver/actions/custom_profile_fields.py:
```diff
@@ -141,6 +191,30 @@ def try_update_realm_custom_profile_field(
+                "hint": field.hint,
+                "required": field.required,
+                "editable_by_user": field.editable_by_user,
+                "display_in_profile_summary": field.display_in_profile_summary,
```
Other way here can be to add a different entry for each property being changed by adding a `property` field to `extra_data` so that we do not need to store the values that haven't been changed. But am not sure what would be better

sahil839 reviewed (commented):
Posted some comments. You can make the changes for tests now and can wait for Tim's feedback before making changes for other comments as I am not sure about them.

sahil839 commented on zerver/tests/test_audit_log.py:
```diff
@@ -568,6 +575,161 @@ def test_set_realm_authentication_methods(self) -> None:
+        self.assertEqual(audit_log.acting_user, user)
+        self.assertEqual(audit_log.modified_user, user)
+        self.assertEqual(audit_log.extra_data["field_id"], field.id)
+        self.assertEqual(audit_log.extra_data[RealmAuditLog.NEW_VALUE], "New Updated Bio")
```
Would be good to check `extra_data` like this `self.assertEqual(audit_log.extra_data, {"field_id": field.id, RealmAuditLog.NEW_VALUE: "New updated Bio", RealmAuditLog.OLD_VALUE: "Original Bio"}` so that we can be sure there are no unexpected fields in the object. You can also use `expected_extra_data` variable if the object is big enough.

sahil839 commented on zerver/tests/test_audit_log.py:
```diff
@@ -568,6 +575,165 @@ def test_set_realm_authentication_methods(self) -> None:
+            RealmAuditLog.NEW_VALUE: "New Updated Bio",
+            RealmAuditLog.OLD_VALUE: "Original Bio",
+        }
+        self.assertEqual(audit_log.extra_data, expected_extra_data)
```
Please add a test for the case when user removes the value as well.

RajVishh commented:
> Some of the previously requested changes are not done for all the tests.
> 
> Also, I missed couple of places where the `RealmAuditLog` entries are not being created right now, like when creating a default external account type field and while reordering the custom profile fields.

I’ve updated the remaining tests to assert full extra_data, added coverage for removing a custom profile field value, and implemented the missing RealmAuditLog entries for default external account field creation and custom profile field reordering.
please take a look when you have a moment.

sahil839 commented on zerver/actions/custom_profile_fields.py:
```diff
@@ -45,6 +53,19 @@ def try_add_realm_default_custom_profile_field(
+
+    RealmAuditLog.objects.create(
+        realm=realm,
+        acting_user=None,  # system action
```
And other than that it is only called from `populate_db` where you are already passing `iago` as acting user when creating other profile fields, so that can be done here as well.

sahil839 commented on zerver/actions/custom_profile_fields.py:
```diff
@@ -156,6 +248,18 @@ def try_reorder_realm_custom_profile_fields(realm: Realm, order: Iterable[int])
+        event_time=timezone_now(),
+        extra_data={
+            RealmAuditLog.OLD_VALUE: old_order,
+            RealmAuditLog.NEW_VALUE: list(order),
```
This does not give sufficient details on what is updated. I think you can add `property` field and set it to "order".

sahil839 replied:
Maybe we might need a separate `event_type` because this is not related to a single custom profile field, but I am not sure so that can be decided later when Tim reviews.

whilstsomebody commented:
@RajVishh , can I continue with the PR to complete the remaining work on tnis??

abhay5-6 commented:
Hi @RajVishh! Thank you for the great initial work on this. Since it looks like this PR has been inactive for a while and has some merge conflicts, I’m going to go ahead and pick up where you left off .

I'll be starting a new branch to resolve the merge conflicts and implement the remaining feedback from @sahil839 and @timabbott. I'll make sure to link back to this PR for context. Thanks again for laying the groundwork.
